src/dune_rules/ctypes_rules.ml

```diff
@@ -148,38 +148,63 @@ let write_function_gen_script ~include_headers ~sctx ~dir ~name
   let script = function_gen_gen ~include_headers ~function_description_module in
   Super_context.add_rule ~loc:Loc.none sctx ~dir (Build.write_file path script)
 
-let rule ?stdout_to ?(action_args=[]) ?(targets=[]) ~exe ~sctx ~dir () =
+let rule ?(deps=[]) ?stdout_to ?(args=[]) ?(targets=[]) ~exe ~sctx ~dir () =
   let build =
-    let targets = List.map targets ~f:(Path.Build.relative dir) in
-    let exe = Ok (Path.build (Path.Build.relative dir exe)) in
+    let exe =
+      match exe with
+      | `relative exe -> Ok (Path.build (Path.Build.relative dir exe))
+      | `unresolved exe -> Super_context.resolve_program ~loc:None ~dir sctx exe
+    in
     let args =
+      let targets = List.map targets ~f:(Path.Build.relative dir) in
+      let deps =
+        List.map deps ~f:(Path.relative (Path.build dir))
+        |> Dep.Set.of_files
+      in
       let open Command.Args in
       [ Hidden_targets targets
-      ; As action_args ]
+      ; Hidden_deps deps
+      ; As args ]
     in
     let stdout_to = Option.map stdout_to ~f:(Path.Build.relative dir) in
     Command.run exe ~dir:(Path.build dir) ?stdout_to args
   in
   Super_context.add_rule sctx ~dir build
 
-let rule_shell_action ~deps ~targets ~action_args ~sctx ~dir () =
-  let build =
-    let sh_path, sh_arg = Utils.system_shell_exn ~needed_to:"build ctypes" in
-    let targets = List.map targets ~f:(Path.Build.relative dir) in
-    let deps =
-      List.map deps ~f:(Path.relative (Path.build dir))
-      |> Dep.Set.of_files
-    in
-    let exe = Ok sh_path in
-    let args =
-      let open Command.Args in
-      [ As (sh_arg :: action_args)
-      ; Hidden_targets targets
-      ; Hidden_deps deps ]
+let build_c_program ~sctx ~dir ~source_files ~scope ~cflags_txt ~output () =
```

Comment:
What about creating a general purpose `Foreign_rules.build_exe` function? It's not required, but it would be nice.